LoopKit/HealthKitSampleStore.swift

```diff
@@ -221,51 +187,59 @@ public class HealthKitSampleStore {
+            let queryAnchor = self.queryAnchor
+            
+            self.log.default("%@ notified with changes. Fetching from: %{public}@", query, queryAnchor.map(String.init(describing:)) ?? "0")
+            
+            let semaphore = DispatchSemaphore(value: 0)
 
-        let anchoredObjectQuery = createAnchoredObjectQuery(sampleType, query.predicate, queryAnchor, HKObjectQueryNoLimit) {  [weak self] (query, newSamples, deletedSamples, anchor, error) in
-            self?.anchoredObjectQueryResultsHandler(query: query, newSamples: newSamples, deletedSamples: deletedSamples, anchor: anchor, error: error)
+            let anchoredObjectQuery = self.createAnchoredObjectQuery(self.sampleType, query.predicate, queryAnchor, HKObjectQueryNoLimit) { [weak self] (query, newSamples, deletedSamples, anchor, error) in
+                
+                if let self = self {
+                    self.anchoredObjectQueryResultsHandler(query: query, newSamples: newSamples, deletedSamples: deletedSamples, anchor: anchor, error: error) {
+                        completionHandler()
+                        semaphore.signal()
+                    }
+                } else {
+                    completionHandler()
+                    semaphore.signal()
+                }
+            }
+            self.healthStore.execute(anchoredObjectQuery)
+            
+            semaphore.wait()
         }
-
-        healthStore.execute(anchoredObjectQuery)
     }
-
-    private final func anchoredObjectQueryResultsHandler(query: HKAnchoredObjectQuery, newSamples: [HKSample]?, deletedSamples: [HKDeletedObject]?, anchor: HKQueryAnchor?, error: Error?) {
-        
+    
+    private func anchoredObjectQueryResultsHandler(query: HKAnchoredObjectQuery, newSamples: [HKSample]?, deletedSamples: [HKDeletedObject]?, anchor: HKQueryAnchor?, error: Error?, completion: @escaping () -> Void) {
         if let error = error {
-            log.error("Error from anchoredObjectQuery: anchor: %{public}@ error: %{public}@", String(describing: anchor), String(describing: error))
-            // We continue here instead of returning, observerQueryCompletionHandler must be called below.
-        } else {
-            log.default("anchoredObjectQuery.resultsHandler: new: %{public}d deleted: %{public}d anchor: %{public}@", newSamples?.count ?? 0, deletedSamples?.count ?? 0, String(describing: anchor))
+            self.log.error("Error from anchoredObjectQuery: anchor: %{public}@ error: %{public}@", String(describing: anchor), String(describing: error))
+            completion()
+            return
         }
-
+        
+        self.log.default("anchoredObjectQuery.resultsHandler: new: %{public}d deleted: %{public}d anchor: %{public}@", newSamples?.count ?? 0, deletedSamples?.count ?? 0, String(describing: anchor))
+        
         guard let anchor = anchor else {
-            callAndUpdateObserverQueryCompletionHandler(newValue: nil)
+            completion()
```

Comment:
[nit] This is really an unexpected state, right? That is, an anchor query that is returning a *new* anchor that is `nil`. Consider an error log statement here.